src/goto-instrument/contracts/dynamic-frames/dfcc_dsl_contract_functions.cpp

```diff
@@ -0,0 +1,572 @@
+  {
+    const auto &funcall = to_side_effect_expr_function_call(target);
+    if(can_cast_expr<symbol_exprt>(funcall.function()))
+    {
+      const auto &sym_expr = to_symbol_expr(funcall.function());
+      const auto &ident = sym_expr.get_identifier();
+      auto hook = library.get_hook(ident);
+      if(hook.has_value())
+      {
+        // if `ident` is one of the built-ins
+        //
+        // `__CPROVER_object_from`
+        // `__CPROVER_object_upto`
+        // `__CPROVER_object_whole`
+        // `__CPROVER_assignable`
+        //
+        // generate a call to the built-in
+        //
+        // ```
+        // CALL ident(list, args);
+        // ```
+        code_function_callt code_function_call(sym_expr);
+        auto &arguments = code_function_call.arguments();
+        for(auto &arg : funcall.arguments())
+          arguments.emplace_back(arg);
+
+        dest.add(goto_programt::make_function_call(
+          code_function_call, source_location));
+      }
+      else
+      {
+        // Calls to user-defined __CPROVER_assignable_t functions
+        //
+        // `cond: foo(params)`
+        //
+        // Become
+        //
+        // ```
+        // CALL foo(params);
+        // ```
+        code_function_callt code_function_call(
+          to_symbol_expr(funcall.function()));
+        auto &arguments = code_function_call.arguments();
+        for(auto &arg : funcall.arguments())
+          arguments.emplace_back(arg);
+        dest.add(goto_programt::make_function_call(
+          code_function_call, source_location));
+      }
+    }
+  }
+  else if(is_assignable(target))
+  {
+    // An lvalue `target` becomes
+    //
+    // ```
+    // CALL __CPROVER_assignable(&target, sizeof(target), is_ptr_to_ptr);
+    // ```
+    const auto &size =
+      size_of_expr(target.type(), namespacet(goto_model.symbol_table));
+
+    if(!size.has_value())
+    {
+      log.error().source_location = target.source_location();
+      log.error()
+        << "dfcc_dsl_contract_functionst::encode_assignable_target: no "
+           "definite size for lvalue assigns clause target "
+        << format(target) << messaget::eom;
+      throw 0;
+    }
+    // we have to build the symbol manually because it might not
+    // be present in the symbol table if the user program does not already
+    // use it.
+    code_function_callt code_function_call(
+      symbol_exprt(CPROVER_PREFIX "assignable", empty_typet()));
+    auto &arguments = code_function_call.arguments();
+
+    // ptr
+    arguments.emplace_back(typecast_exprt::conditional_cast(
+      address_of_exprt{target}, pointer_type(empty_typet())));
+
+    // size
+    arguments.emplace_back(
+      typecast_exprt::conditional_cast(size.value(), size_type()));
+
+    // is_ptr_to_ptr
+    exprt is_ptr_to_ptr = false_exprt();
+    if(target.type().id() == ID_pointer)
+      is_ptr_to_ptr = true_exprt();
+
+    arguments.emplace_back(is_ptr_to_ptr);
```

Comment:
I like to name expressions so that it is easier for the reader (my future self) to connect it to the function parameter as declared in the internal additions.

Reply:
Makes sense!